Fix #9052: Disable E2E Test Sharding by Default

U8NWXD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #9052.
2. This PR does the following:

We have been seeing a flake in the end-to-end tests characterized by a
failure of the GAE WSGI server to bind to a port (see #9052). This may
be because of a race condition in how GAE identifies unused ports to
bind (documented in
https://github.com/oppia/oppia/issues/9052#issuecomment-653931365).
Therefore we cannot rely on sharding, so this commit disables it by
default.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

**This PR is a draft because I still need to rerun tests to evaluate flakiness.**

## Test Rerun Log

1. All pass
   
![Screen Shot 2020-07-08 at 19 44 54](https://user-images.githubusercontent.com/19878639/86981124-c2086d80-c153-11ea-8ef2-aeda8911dca4.png)
![Screen Shot 2020-07-08 at 19 45 43](https://user-images.githubusercontent.com/19878639/86981134-c6cd2180-c153-11ea-850a-a2704bb22bac.png)

2. Travis passed; CircleCI had known flakes, so I reran them from failed
![Screen Shot 2020-07-08 at 20 30 55](https://user-images.githubusercontent.com/19878639/86983476-f67f2800-c159-11ea-8c64-4e47cda9fc48.png)

After rerun, CircleCI passed

![Screen Shot 2020-07-09 at 07 35 05](https://user-images.githubusercontent.com/19878639/87035098-d7fb4a00-c1b6-11ea-8a94-96d1ddececc5.png)

3. All passed

![Screen Shot 2020-07-09 at 07 35 53](https://user-images.githubusercontent.com/19878639/87035123-e0ec1b80-c1b6-11ea-9a3b-723e1719176b.png)

![Screen Shot 2020-07-09 at 09 32 13](https://user-images.githubusercontent.com/19878639/87046396-257fb300-c1c7-11ea-8c23-6955cbeab874.png)

4. Travis Passed; CircleCI had unrelated flaked, so restarting from failed

![Screen Shot 2020-07-09 at 09 33 05](https://user-images.githubusercontent.com/19878639/87046443-37615600-c1c7-11ea-88fa-6863804cdaeb.png)

After restarting, CircleCI passed

![Screen Shot 2020-07-09 at 13 07 02](https://user-images.githubusercontent.com/19878639/87069410-21fb2480-c1e5-11ea-9e4d-568fa0a62fcd.png)


5. Travis Passed; unrelated flake in CircleCI. Rerunning CircleCI from failed

![Screen Shot 2020-07-09 at 10 50 30](https://user-images.githubusercontent.com/19878639/87055206-0aff0700-c1d2-11ea-8eb1-5252e3738876.png)

After restarting, CircleCI passed

![Screen Shot 2020-07-09 at 14 09 48](https://user-images.githubusercontent.com/19878639/87075151-e87ae700-c1ed-11ea-9529-b0c1da90429c.png)

U8NWXD committed:
Disable E2E Test Sharding by Default

We have been seeing a flake in the end-to-end tests characterized by a
failure of the GAE WSGI server to bind to a port (see #9052). This may
be because of a race condition in how GAE identifies unused ports to
bind (documented in
https://github.com/oppia/oppia/issues/9052#issuecomment-653931365).
Therefore we cannot rely on sharding, so this commit disables it by
default.

U8NWXD commented:
@DubeySandeep PTAL as codeowner!

U8NWXD committed:
Fix backend test

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -97,7 +97,7 @@
     action='store_true')
 _PARSER.add_argument(
-    '--sharding-instances', type=int, default=3,
+    '--sharding-instances', type=int, default=1,
```
Should we set this only for the tests which are failing with `unable to bind error`?

U8NWXD replied:
done

U8NWXD commented:
~~**Please do not merge this PR**: We still need to rerun the tests to confirm it isn't flaky~~
Edit: I'll mark the PR as ready for review (not draft) once I've rerun the CI

U8NWXD committed:
Revert Disabling E2E Test Sharding

This reverts commits:

* c1efa3f34212c77a65bb94260bb1cfd9561361cd
* e1d7c33a61a1f4f1ea61444868ec639a61873477

We are going to only disable sharding on the problematic suites to avoid
slowing down the tests more than necessary.

U8NWXD committed:
Disable Sharding for Suites with Binding Flake

The Google Appengine code uses
[portpicker](https://github.com/google/python_portpicker), whose readme
says:

> There is a race condition between picking a port and your application
> code binding to it. The use of a port server by all of your test code
> to avoid that problem is recommended on loaded test hosts running many
> tests at a time.

We think this is the cause of the localhost binding flake documented in
issue #9052. As a temporary fix, this commit disables sharding on the
tests where we've seen this flake to eliminate concurrent client
requests as the tests run.

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into disable-ci-sharding

kevinlee12 commented:
@U8NWXD please use Draft PRs if you don't intend PRs to be merged right away (it's like a merge lock by the author). Note in the description that you don't want tests to be cancelled because you're testing them :smile: )

U8NWXD committed:
Disable sharding for library suite

kevinlee12 commented:
@ankita240796 @jameesjohn ptal at the duplicated comments! ^

U8NWXD commented:
@BenHenning @vinitamurthi PTAL as codeowners

jameesjohn commented:
> @ankita240796 @jameesjohn ptal at the duplicated comments! ^

Thanks @kevinlee12.
I'm looking into it.
